File: codes/ruby/yyy_required_features.rb
require "json"

=begin
    1. variable can store dynamic data type and dynamic value, variable can inferred data type from value, value of variable can be reassign with different data type or has option to make variable can store dynamic data type and dynamic value
    ```javascript
    let something = "foo";
    console.log("something:", something);
    something = 123;
    console.log("something:", something);
    something = true;
    console.log("something:", something);
    something = null;
    console.log("something:", something);
    something = [1, 2, 3];
    console.log("something:", something);
    something = { "foo": "bar" };
    console.log("something:", something);
    ```
    ```go
    type Any interface{}
    ```
=end
something = "foo"
print("something: ", JSON.pretty_generate(something, { "indent": " " * 4 }), "\n")
something = 123
print("something: ", JSON.pretty_generate(something, { "indent": " " * 4 }), "\n")
something = true
print("something: ", JSON.pretty_generate(something, { "indent": " " * 4 }), "\n")
something = nil
print("something: ", JSON.pretty_generate(something, { "indent": " " * 4 }), "\n")
something = [1, 2, 3]
print("something: ", JSON.pretty_generate(something, { "indent": " " * 4 }), "\n")
something = { "foo" => "bar" }
print("something: ", JSON.pretty_generate(something, { "indent": " " * 4 }), "\n")

=begin
    2. it is possible to access and modify variables defined outside of the current scope within nested functions, so it is possible to have closure too
    ```javascript
    function getModifiedIndentLevel() {
        let indentLevel = 0;
        function changeIndentLevel() {
            indentLevel += 1;
            if (indentLevel < 5) changeIndentLevel();
            return indentLevel;
        }
        return changeIndentLevel();
    }
    console.log("getModifiedIndentLevel():", getModifiedIndentLevel());
    function createNewGame(initialCredit) {
        let currentCredit = initialCredit;
        console.log("initial credit:", initialCredit);
        return function () {
            currentCredit -= 1;
            if (currentCredit === 0) {
                console.log("not enough credits");
                return;
            }
            console.log(`playing game, ${currentCredit} credit(s) remaining`);
        };
    }
    const playGame = createNewGame(3);
    playGame();
    playGame();
    playGame();
    ```
=end
def get_modified_indent_level()
    indent_level = 0
    change_indent_level = ->() do
        indent_level += 1
        change_indent_level.call() if (indent_level < 5)
        return indent_level
    end
    change_indent_level.call()
end
print("get_modified_indent_level(): ", get_modified_indent_level(), "\n")
def create_new_game(initial_credit)
    current_credit = initial_credit
    print("initial credit: ", initial_credit, "\n")
    return ->() do
        current_credit -= 1
        if (current_credit == 0)
            print("not enough credits", "\n")
            return
        end
        print("playing game, #{current_credit} credit(s) remaining", "\n")
    end
end
play_game = create_new_game(3)
play_game.call()
play_game.call()
play_game.call()

=begin
    3. object/dictionary/associative-array/hash/hashmap/map/unordered-list-key-value-pair-data-structure can store dynamic data type and dynamic value
    ```javascript
    const myObject = {
        "my_string": "foo",
        "my_number": 123,
        "my_bool": true,
        "my_null": null,
        "my_array": [1, 2, 3],
        "my_object": {
            "foo": "bar"
        }
    };
    console.log("myObject:", myObject);
    ```
=end
my_object = {
    "my_string" => "foo",
    "my_number" => 123,
    "my_bool" => true,
    "my_null" => nil,
    "my_array" => [1, 2, 3],
    "my_object" => {
        "foo" => "bar"
    }
}
print("my_object: ", JSON.pretty_generate(my_object, { "indent": " " * 4 }), "\n")

=begin
    4. array/list/slice/ordered-list-data-structure can store dynamic data type and dynamic value
    ```javascript
    const myArray = ["foo", 123, true, null, [1, 2, 3], { "foo": "bar" }];
    console.log("myArray:", myArray);
    ```
=end
my_array = ["foo", 123, true, nil, [1, 2, 3], { "foo" => "bar" }]
print("my_array: ", JSON.pretty_generate(my_array, { "indent": " " * 4 }), "\n")

=begin
    5. support passing functions as arguments to other functions
    ```javascript
    function sayHello(callbackFunction) {
        console.log("hello");
        callbackFunction();
    }
    function sayHowAreYou() {
        console.log("how are you?");
    }
    sayHello(sayHowAreYou);
    sayHello(function () {
        console.log("how are you?");
    });
    ```
=end
def say_hello(callback_function)
    print("hello\n")
    callback_function.call()
end
def say_how_are_you()
    print("how are you?\n")
end
say_hello(method(:say_how_are_you))
say_hello(->() do
    print("how are you?", "\n")
end)

=begin
    6. support returning functions as values from other functions
    ```javascript
    function multiply(a) {
        return function (b) {
            return (a * b);
        };
    }
    const multiplyBy2 = multiply(2);
    const multiplyBy2Result = multiplyBy2(10);
    console.log("multiplyBy2Result:", multiplyBy2Result);
    ```
=end
multiply = ->(a) do
    return ->(b) do
        return (a * b)
    end
end
multiply_by2 = multiply.call(2)
multiply_by2_result = multiply_by2.call(10)
print("multiply_by2_result: ", multiply_by2_result, "\n")

=begin
    7. support assigning functions to variables
    ```javascript
    const getRectangleAreaV1 = function (rectangleWidth, rectangleLength) {
        return (rectangleWidth * rectangleLength);
    };
    console.log(`getRectangleAreaV1(7, 5): ${getRectangleAreaV1(7, 5)}`);
    const getRectangleAreaV2 = (rectangleWidth, rectangleLength) => {
        return (rectangleWidth * rectangleLength);
    };
    console.log(`getRectangleAreaV2(7, 5): ${getRectangleAreaV2(7, 5)}`);
    const getRectangleAreaV3 = (rectangleWidth, rectangleLength) => (rectangleWidth * rectangleLength);
    console.log(`getRectangleAreaV3(7, 5): ${getRectangleAreaV3(7, 5)}`);
    ```
=end
get_rectangle_area_v1 = Proc.new { |rectangle_width, rectangle_length| (rectangle_width * rectangle_length) }
print("get_rectangle_area_v1.call(7, 5): ", get_rectangle_area_v1.call(7, 5), "\n")
get_rectangle_area_v2 = proc { |rectangle_width, rectangle_length| (rectangle_width * rectangle_length) }
print("get_rectangle_area_v2.call(7, 5): ", get_rectangle_area_v2.call(7, 5), "\n")
get_rectangle_area_v3 = Proc.new do |rectangle_width, rectangle_length|
    (rectangle_width * rectangle_length)
end
print("get_rectangle_area_v3.call(7, 5): ", get_rectangle_area_v3.call(7, 5), "\n")
get_rectangle_area_v4 = proc do |rectangle_width, rectangle_length|
    (rectangle_width * rectangle_length)
end
print("get_rectangle_area_v4.call(7, 5): ", get_rectangle_area_v4.call(7, 5), "\n")
get_rectangle_area_v5 = lambda { |rectangle_width, rectangle_length| (rectangle_width * rectangle_length) }
print("get_rectangle_area_v5.call(7, 5): ", get_rectangle_area_v5.call(7, 5), "\n")
get_rectangle_area_v6 = ->(rectangle_width, rectangle_length) { (rectangle_width * rectangle_length) }
print("get_rectangle_area_v6.call(7, 5): ", get_rectangle_area_v6.call(7, 5), "\n")
get_rectangle_area_v7 = lambda do |rectangle_width, rectangle_length|
    return (rectangle_width * rectangle_length)
end
print("get_rectangle_area_v7.call(7, 5): ", get_rectangle_area_v7.call(7, 5), "\n")
get_rectangle_area_v8 = ->(rectangle_width, rectangle_length) do
    return (rectangle_width * rectangle_length)
end
print("get_rectangle_area_v8.call(7, 5): ", get_rectangle_area_v8.call(7, 5), "\n")

=begin
    8. support storing functions in data structures like array/list/slice/ordered-list-data-structure or object/dictionary/associative-array/hash/hashmap/map/unordered-list-key-value-pair-data-structure
    ```javascript
    const myArray2 = [
        function (a, b) {
            return (a * b);
        },
        "foo",
        123,
        true,
        null,
        [1, 2, 3],
        { "foo": "bar" }
    ];
    console.log("myArray2[0](7, 5):", myArray2[0](7, 5));

    const myObject2 = {
        "my_function": function (a, b) {
            return (a * b);
        },
        "my_string": "foo",
        "my_number": 123,
        "my_bool": true,
        "my_null": null,
        "my_array": [1, 2, 3],
        "my_object": {
            "foo": "bar"
        }
    };
    console.log("myObject2["my_function"](7, 5):", myObject2["my_function"](7, 5));
    ```
=end
my_array2 = [
    ->(a, b) do
        return (a * b)
    end,
    "foo",
    123,
    true,
    nil,
    [1, 2, 3],
    { "foo" => "bar" }
]
print("myArray2[0](7, 5): ", my_array2[0].call(7, 5), "\n")

my_object2 = {
    "my_function" => ->(a, b) do
        return (a * b)
    end,
    "my_string" => "foo",
    "my_number" => 123,
    "my_bool" => true,
    "my_null" => nil,
    "my_array" => [1, 2, 3],
    "my_object" => {
        "foo" => "bar"
    }
}
print("myObject2[\"my_function\"](7, 5): ", my_object2["my_function"].call(7, 5), "\n")

File: codes/php/yyy_required_features.php
<?php

$js_like_type = ["Null" => "Null", "Boolean" => "Boolean", "String" => "String", "Numeric" => "Numeric", "Object" => "Object", "Array" => "Array", "Function" => "Function"];

$array_reduce_v2 = function ($callback_function, $an_array, $initial_value) {
    // JavaScript-like Array.reduce() function $array_reduce_v2
    $result = $initial_value;
    $array_item_index = 0;
    foreach ($an_array as $object_key => $object_value) {
        $result = $callback_function($result, $object_value, $array_item_index, $an_array);
        $array_item_index += 1;
    }
    return $result;
};

$array_every = function ($callback_function, $an_array) {
    // JavaScript-like Array.every() function $array_every_v4
    $array_item_index = 0;
    foreach ($an_array as $object_key => $object_value) {
        if ($callback_function($object_value, $array_item_index, $an_array) === false) return false;
        $array_item_index += 1;
    }
    return true;
};

$is_like_js_null = fn($anything) => (((gettype($anything) === "null") || (gettype($anything) === "NULL")) && (is_null($anything) === true) && (isset($anything) === false));

$is_like_js_boolean = fn($anything) => (((gettype($anything) === "boolean") || (gettype($anything) === "bool")) && (is_bool($anything) === true) && (($anything === true) || ($anything === false)));

$is_like_js_string = fn($anything) => ((gettype($anything) === "string") && (is_string($anything) === true));

$is_like_js_numeric = fn($anything) => (((gettype($anything) === "integer") || (gettype($anything) === "int") || (gettype($anything) === "float") || (gettype($anything) === "double")) && ((is_int($anything) === true) || (is_integer($anything) === true) || (is_float($anything) === true) || (is_double($anything) === true)) && (is_numeric($anything) === true));

$is_like_js_array = fn($anything) => ((gettype($anything) === "array") && (is_array($anything) === true));

$is_like_js_object = fn($anything) => (($is_like_js_array($anything) === false) ? false : ($array_every((fn($object_value) => ($is_like_js_string($object_value) === true)), array_keys($anything))));

$is_like_js_function = fn($anything) => (is_callable($anything) === true);

$get_type = fn($anything) => (($is_like_js_null($anything) === true) ? ($js_like_type["Null"]) : (($is_like_js_boolean($anything) === true) ? ($js_like_type["Boolean"]) : ((($is_like_js_string($anything) === true) ? ($js_like_type["String"]) : (($is_like_js_numeric($anything) === true) ? ($js_like_type["Numeric"]) : ((($is_like_js_object($anything) === true) ? ($js_like_type["Object"]) : (($is_like_js_array($anything) === true) ? ($js_like_type["Array"]) : (($is_like_js_function($anything) === true) ? ($js_like_type["Function"]) : ('"' . gettype($anything) . '"'))))))))));

$optional_chaining = fn($anything, ...$array_index_or_object_key_or_function_argument_array) => (($get_type($anything) === $js_like_type["Function"]) ? ($anything(...$array_index_or_object_key_or_function_argument_array)) : (((($get_type($anything) !== $js_like_type["Object"]) && ($get_type($anything) !== $js_like_type["Array"])) || (count($array_index_or_object_key_or_function_argument_array) === 0)) ? ($anything) : ($array_reduce_v2((fn($current_result, $current_item) => ((($get_type($current_result) === $js_like_type["Null"]) && ($get_type($anything) === $js_like_type["Object"]) && ($get_type($current_item) === $js_like_type["String"])) ? (@$anything[((string) $current_item)]) : ((($get_type($current_result) === $js_like_type["Null"]) && ($get_type($anything) === $js_like_type["Array"]) && ($get_type($current_item) === $js_like_type["Numeric"]) && (((int) $current_item) >= 0) && (count($anything) > ((int) $current_item))) ? (@$anything[((int) $current_item)]) : ((($get_type($current_result) === $js_like_type["Object"]) && ($get_type($current_item) === $js_like_type["String"])) ? (@$current_result[((string) $current_item)]) : ((($get_type($current_result) === $js_like_type["Array"]) && ($get_type($current_item) === $js_like_type["Numeric"]) && (((int) $current_item) >= 0) && (count($current_result) > ((int) $current_item))) ? (@$current_result[((int) $current_item)]) : null))))), $array_index_or_object_key_or_function_argument_array, null))));

$pipe = function (...$rest_arguments) use ($js_like_type, $get_type, $array_reduce_v2) {
    $pipe_last_result = null;
    $pipe_result = $array_reduce_v2((function ($current_result, $current_argument) use (&$pipe_last_result, $js_like_type, $get_type) {
        $pipe_last_result = $current_result;
        return (($get_type($current_result) === $js_like_type["Null"]) ? ($current_argument) : (($get_type($current_argument) === $js_like_type["Function"]) ? ($current_argument($current_result)) : (null)));
    }), $rest_arguments, null);
    if ($get_type($pipe_result) === $js_like_type["Function"]) return $pipe_result($pipe_last_result);
    return $pipe_result;
};

$json_stringify = function ($anything, $pretty = false) use ($js_like_type, $get_type) {
    // $json_stringify_v4
    $indent = str_repeat(" ", 4);
    $indent_level = 0;
    $json_stringify_inner = function ($anything_inner) use ($js_like_type, $get_type, $pretty, $indent, &$indent_level, &$json_stringify_inner) {
        if ($get_type($anything_inner) === $js_like_type["Null"]) return "null";
        if ($get_type($anything_inner) === $js_like_type["String"]) return '"' . $anything_inner . '"';
        if ($get_type($anything_inner) === $js_like_type["Numeric"]) return ((string) $anything_inner);
        if (($get_type($anything_inner) === $js_like_type["Boolean"]) && ($anything_inner === true)) return "true";
        if (($get_type($anything_inner) === $js_like_type["Boolean"]) && ($anything_inner === false)) return "false";
        if ($get_type($anything_inner) === $js_like_type["Object"]) {
            if (count($anything_inner) === 0) return "{}";
            $indent_level += 1;
            $result = (($pretty === true) ? ("{\n" . str_repeat($indent, $indent_level)) : "{ ");
            $object_entry_index = 0;
            foreach ($anything_inner as $object_key => $object_value) {
                $result .= ('"' . $object_key . '": ' . $json_stringify_inner($object_value));
                if (($object_entry_index + 1) !== count($anything_inner)) {
                    $result .= (($pretty === true) ? (",\n" . str_repeat($indent, $indent_level)) : ", ");
                }
                $object_entry_index += 1;
            }
            $indent_level -= 1;
            $result .= (($pretty === true) ? ("\n" . str_repeat($indent, $indent_level) . "}") : " }");
            return $result;
        }
        if ($get_type($anything_inner) === $js_like_type["Array"]) {
            if (count($anything_inner) === 0) return "[]";
            $indent_level += 1;
            $result = (($pretty === true) ? ("[\n" . str_repeat($indent, $indent_level)) : "[");
            $array_item_index = 0;
            foreach ($anything_inner as $object_key => $object_value) {
                $result .= $json_stringify_inner($object_value);
                if (($array_item_index + 1) !== count($anything_inner)) {
                    $result .= (($pretty === true) ? (",\n" . str_repeat($indent, $indent_level)) : ", ");
                }
                $array_item_index += 1;
            }
            $indent_level -= 1;
            $result .= (($pretty === true) ? ("\n" . str_repeat($indent, $indent_level) . "]") : "]");
            return $result;
        }
        if ($get_type($anything_inner) === $js_like_type["Function"]) return '"[object Function]"';
        return '"' . $get_type($anything_inner) . '"';
    };
    return $json_stringify_inner($anything);
};

$string_interpolation = fn(...$rest_arguments) => $array_reduce_v2((fn($current_result, $current_argument) => ($current_result . (($get_type($current_argument) === $js_like_type["String"]) ? ($current_argument) : ((($get_type($current_argument) === $js_like_type["Array"]) && (count($current_argument) === 1)) ? ($json_stringify($optional_chaining($current_argument, 0))) : ($json_stringify($current_argument)))))), $rest_arguments, "");

$console_log = function (...$rest_arguments) use ($string_interpolation) {
    echo $string_interpolation(...$rest_arguments) . "\n";
};

/*
1. variable can store dynamic data type and dynamic value, variable can inferred data type from value, value of variable can be reassign with different data type or has option to make variable can store dynamic data type and dynamic value
```javascript
let something = "foo";
console.log(`something: ${something}`);
something = 123;
console.log(`something: ${something}`);
something = true;
console.log(`something: ${something}`);
something = null;
console.log(`something: ${something}`);
something = [1, 2, 3];
console.log(`something: ${something}`);
something = { "foo": "bar" };
console.log(`something: ${something}`);
```
```go
type Any interface{}
```
*/
$something = "foo";
$console_log($string_interpolation("something: ", $json_stringify($something, pretty: true)));
$something = 123;
$console_log($string_interpolation("something: ", $json_stringify($something, pretty: true)));
$something = true;
$console_log($string_interpolation("something: ", $json_stringify($something, pretty: true)));
$something = null;
$console_log($string_interpolation("something: ", $json_stringify($something, pretty: true)));
$something = [1, 2, 3];
$console_log($string_interpolation("something: ", $json_stringify($something, pretty: true)));
$something = ["foo" => "bar"];
$console_log($string_interpolation("something: ", $json_stringify($something, pretty: true)));

/*
2. it is possible to access and modify variables defined outside of the current scope within nested functions, so it is possible to have closure too
```javascript
function getModifiedIndentLevel() {
    let indentLevel = 0;
    function changeIndentLevel() {
        indentLevel += 1;
        if (indentLevel < 5) changeIndentLevel();
        return indentLevel;
    }
    return changeIndentLevel();
}
console.log(`getModifiedIndentLevel(): ${getModifiedIndentLevel()}`);
function createNewGame(initialCredit) {
    let currentCredit = initialCredit;
    console.log(`initial credit: ${initialCredit}`);
    return function () {
        currentCredit -= 1;
        if (currentCredit === 0) {
            console.log("not enough credits");
            return;
        }
        console.log(`playing game, ${currentCredit} credit(s) remaining`);
    };
}
const playGame = createNewGame(3);
playGame();
playGame();
playGame();
```
*/
$get_modified_indent_level = function () {
    $indent_level = 0;
    $change_indent_level = function() use (&$indent_level, &$change_indent_level) {
        $indent_level += 1;
        if ($indent_level < 5) $change_indent_level();
        return $indent_level;
    };
    return $change_indent_level();
};
$console_log($string_interpolation("get_modified_indent_level(): ", [$get_modified_indent_level()]));
$create_new_game = function ($initial_credit) use ($console_log, $string_interpolation) {
    $current_credit = $initial_credit;
    $console_log($string_interpolation("initial_credit: ", [$initial_credit]));
    return function () use ($console_log, $string_interpolation, &$current_credit) {
        $current_credit -= 1;
        if ($current_credit === 0) {
            $console_log("not enough credits");
            return;
        }
        $console_log($string_interpolation("playing game, ", [$current_credit], " credit(s) remaining"));
    };
};
$play_game = $create_new_game(3);
$play_game();
$play_game();
$play_game();

/*
3. object/dictionary/associative-array/hash/hashmap/map/unordered-list-key-value-pair-data-structure can store dynamic data type and dynamic value
```javascript
const myObject = {
    "my_string": "foo",
    "my_number": 123,
    "my_bool": true,
    "my_null": null,
    "my_array": [1, 2, 3],
    "my_object": {
        "foo": "bar"
    }
};
console.log(`myObject: ${myObject}`);
```
*/
$my_object = [
    "my_string" => "foo",
    "my_number" => 123,
    "my_bool" => true,
    "my_null" => null,
    "my_array" => [1, 2, 3],
    "my_object" => [
        "foo" => "bar"
    ]
];
$console_log($string_interpolation("my_object: ", $json_stringify($my_object, pretty: true)));

/*
4. array/list/slice/ordered-list-data-structure can store dynamic data type and dynamic value
```javascript
const myArray = ["foo", 123, true, null, [1, 2, 3], { "foo": "bar" }];
console.log(`myArray: ${myArray}`);
```
*/
$my_array = ["foo", 123, true, null, [1, 2, 3], ["foo" => "bar"]];
$console_log($string_interpolation("my_array: ", $json_stringify($my_array, pretty: true)));

/*
5. support passing functions as arguments to other functions
```javascript
function sayHello(callbackFunction) {
    console.log("hello");
    callbackFunction();
}
function sayHowAreYou() {
    console.log("how are you?");
}
sayHello(sayHowAreYou);
sayHello(function () {
    console.log("how are you?");
});
```
*/
function say_hello($callback_function) {
    global $console_log;
    $console_log("hello");
    $callback_function();
};
$say_hello_v2 = function ($callback_function) use ($console_log) {
    $console_log("hello");
    $callback_function();
};
function say_how_are_you() {
    global $console_log;
    $console_log("how are you?");
};
$say_how_are_you_v2 = function () use ($console_log) {
    $console_log("how are you?");
};
say_hello("say_how_are_you"); // 🤮
say_hello($say_how_are_you_v2);
$say_hello_v2($say_how_are_you_v2);
say_hello((fn() => (say_how_are_you())));
say_hello(function() use ($console_log) {
    $console_log("how are you?");
});
$say_hello_v2((fn() => (say_how_are_you())));
$say_hello_v2(function() use ($console_log) {
    $console_log("how are you?");
});

/*
6. support returning functions as values from other functions
```javascript
function multiply(a) {
    return function (b) {
        return (a * b);
    };
}
const multiplyBy2 = multiply(2);
const multiplyBy2Result = multiplyBy2(10);
console.log(`multiplyBy2(10): ${multiplyBy2Result}`);
```
*/
$multiply = function ($a) {
    return function ($b) use ($a) {
        return ($a * $b);
    };
};
$multiply_by2 = $multiply(2);
$multiply_by2_result = $multiply_by2(10);
$console_log($string_interpolation("multiplyBy2(10): ", $multiply_by2_result));

/*
7. support assigning functions to variables
```javascript
const getRectangleAreaV1 = function (rectangleWidth, rectangleLength) {
    return (rectangleWidth * rectangleLength);
};
console.log(`getRectangleAreaV1(7, 5): ${getRectangleAreaV1(7, 5)}`);
const getRectangleAreaV2 = (rectangleWidth, rectangleLength) => {
    return (rectangleWidth * rectangleLength);
};
console.log(`getRectangleAreaV2(7, 5): ${getRectangleAreaV2(7, 5)}`);
const getRectangleAreaV3 = (rectangleWidth, rectangleLength) => (rectangleWidth * rectangleLength);
console.log(`getRectangleAreaV3(7, 5): ${getRectangleAreaV3(7, 5)}`);
```
*/
$get_rectangle_area_v1 = function ($rectangle_width, $rectangle_length) {
    return ($rectangle_width * $rectangle_length);
};
$console_log($string_interpolation("get_rectangle_area_v1(7, 5): ", $get_rectangle_area_v1(7, 5)));
$get_rectangle_area_v2 = fn($rectangle_width, $rectangle_length) => ($rectangle_width * $rectangle_length);
$console_log($string_interpolation("get_rectangle_area_v2(7, 5): ", $get_rectangle_area_v2(7, 5)));

/*
8. support storing functions in data structures like array/list/slice/ordered-list-data-structure or object/dictionary/associative-array/hash/hashmap/map/unordered-list-key-value-pair-data-structure
```javascript
const myArray2 = [
    function (a, b) {
        return (a * b);
    },
    "foo",
    123,
    true,
    null,
    [1, 2, 3],
    { "foo": "bar" }
];
console.log(`myArray2[0](7, 5): ${myArray2[0](7, 5)}`);
const myObject2 = {
    "my_function": function (a, b) {
        return (a * b);
    },
    "my_string": "foo",
    "my_number": 123,
    "my_bool": true,
    "my_null": null,
    "my_array": [1, 2, 3],
    "my_object": {
        "foo": "bar"
    }
};
console.log(`myObject2["my_function"](7, 5): ${myObject2["my_function"](7, 5)}`);
```
*/
$my_array2 = [
    function ($a, $b) {
        return ($a * $b);
    },
    "foo",
    123,
    true,
    null,
    [1, 2, 3],
    ["foo" => "bar"]
];
$console_log($string_interpolation("myArray2[0](7, 5): ", $optional_chaining($optional_chaining($my_array2, 0), 7, 5)));
$console_log($string_interpolation("myArray2[0](7, 5): ", $pipe($optional_chaining($my_array2, 0), (fn($_) => $optional_chaining($_, 7, 5)))));
$my_object2 = [
    "my_function" => function ($a, $b) {
        return ($a * $b);
    },
    "my_string" => "foo",
    "my_number" => 123,
    "my_bool" => true,
    "my_null" => null,
    "my_array" => [1, 2, 3],
    "my_object" => [
        "foo" => "bar"
    ]
];
$console_log($string_interpolation('myObject2["my_function"](7, 5): ', $optional_chaining($optional_chaining($my_object2, "my_function"), 7, 5)));
$console_log($string_interpolation('myObject2["my_function"](7, 5): ', $pipe($optional_chaining($my_object2, "my_function"), (fn($_) => $optional_chaining($_, 7, 5)))));
